### PR TITLE
rust: update to 1.73.0

### PR DESCRIPTION
**Pull request by oskarirauta**
changelog at https://github.com/rust-lang/rust/releases/tag/1.73.0

Maintainer: @lu-zero 
Compile tested: x86_64, latest git
Run tested: x86_64, latest git

### CODE REVIEW

**Comment by oskarirauta** [comment]
Also preparation to conversation started [here](https://github.com/openwrt/packages/pull/22339#issuecomment-1753948111) about musl host support, but this just updates to 1.73.0.

@Ansuel 
rust has been long supported with musl(host) with various repositories, such as alpine. Also if openwrt is supported as a build host, by using openwrt/openwrt#12667 and as it seems, with gentoo as well now that 1.73.0 is out, by just patching it to use newer libc from gentoo at [here](https://gitweb.gentoo.org/repo/gentoo.git/tree/dev-lang/rust/files/1.72.0-bump-libc-deps-to-0.2.146.patch?id=515b5920046117355d88b3494c74da269ce9b30a)

support provided by alpine is a bit too complex, I was never able to adopt it to openwrt packages. With 1.73.0 and libc patch, it worked like a charm with `--set=target.$(RUSTC_TARGET_ARCH).crt-static=false` in `TARGET_CONFIGURE_ARGS` - using that with glibc does not hurt, but I think it would be nicer to retain it as it is and use it only if build host is musl.

But this is just conversation about topic. Hopefully idea gets some hope, but let's update rust first.

**Comment by oskarirauta** [comment]
converted to draft for the moment to do some local checking. I'll get back to this..

**Comment by lu-zero** [comment]
> support provided by alpine is a bit too complex, I was never able to adopt it to openwrt packages. With 1.73.0 and libc patch, it worked like a charm with `--set=target.$(RUSTC_TARGET_ARCH).crt-static=false` in `TARGET_CONFIGURE_ARGS` - using that with glibc does not hurt, but I think it would be nicer to retain it as it is and use it only if build host is musl.

Doesn't hurt and since OpenWrt favours shared libraries it is future proof for when we'll have to consider other libcs.

File: lang/rust/patches/0002-rustc-bootstrap-cache.patch [hunk 1]
```diff
@@ -23,10 +23,10 @@
...
          if !cache_dir.exists() {
              t!(fs::create_dir_all(&cache_dir));
-@@ -627,7 +630,10 @@ download-rustc = false
+@@ -647,7 +650,10 @@ download-rustc = false
          let llvm_assertions = self.llvm_assertions;
  
-         let cache_prefix = format!("llvm-{}-{}", llvm_sha, llvm_assertions);
+         let cache_prefix = format!("llvm-{llvm_sha}-{llvm_assertions}");
 -        let cache_dst = self.out.join("cache");
 +        let cache_dst = match env::var_os("OPENWRT_RUSTC_BOOTSTRAP_CACHE") {
```
Comment (lu-zero) [inline lang/rust/patches/0002-rustc-bootstrap-cache.patch:31]: Probably we could float it towards upstream using a less os-specific name

**Review by lu-zero — APPROVED** [review]
Looks fine, thank you for updating it :)

**Comment by oskarirauta** [comment]
> Doesn't hurt and since OpenWrt favours shared libraries it is future proof for when we'll have to consider other libcs.

Should I then add it to this already?

File: lang/rust/patches/0002-rustc-bootstrap-cache.patch (hunk 1 above)
Comment (oskarirauta) [reply lang/rust/patches/0002-rustc-bootstrap-cache.patch:31]: I'll put there whatever you want- that is just refreshed patch that appeared in previous package release update (1.72.0-3) without me doing any actual changes.

**Comment by lu-zero** [comment]
I'm fine with folding it in :) (and also your change for crt-static)

**Comment by oskarirauta** [comment]
Okay, thanks.. I'll make the necessary adjustment to PR..

**Comment by oskarirauta** [comment]
@lu-zero 

Where are we at with that os branding topic within patch 0002-rustc-bootstrap-cache.patch?

**Comment by jefferyto** [comment]
Re: 0002-rustc-bootstrap-cache.patch - the bootstrap already "caches" downloaded files in the build directory. I imagine this is fine for normal users but for us the build directory is not persisted between builds. I'm not sure upstream would be interested in adding more customization (like an environment variable) for where to save this cache, but then again I never asked.

**Comment by lu-zero** [comment]
I think it is a good idea overall, sometimes you may have more than a tree around even in other scenarios :)

**Comment by oskarirauta** [comment]
So, how do we proceed?

**Comment by jefferyto** [comment]
> I'd land this set, anybody willing to change the env name to `RUSTC_BOOTSTRAP_CACHE` and send upstream for the future?

I opened https://github.com/rust-lang/rust/pull/116697. I think the patch here can stay as it is until if/when the change is merged upstream.

**Comment by oskarirauta** [comment]
Though I thought that I don't participate in decision about it, I have to say that I agree; maybe it's not topic of this PR. Do good.

**Pull request merged**